index.d.ts

```diff
@@ -0,0 +1,490 @@
+export function gte_<A>(x: undefined): (y: undefined) => boolean;
+export function gte_<A>(x: boolean, y: boolean): boolean;
+export function gte_<A>(x: boolean): (y: boolean) => boolean;
+export function gte_<A>(x: number, y: number): boolean;
+export function gte_<A>(x: number): (y: number) => boolean;
+export function gte_<A>(x: Date, y: Date): boolean;
+export function gte_<A>(x: Date): (y: Date) => boolean;
+export function gte_<A>(x: string, y: string): boolean;
+export function gte_<A>(x: string): (y: string) => boolean;
+export function gte_<A>(x: Array<A>, y: Array<A>): boolean;
+export function gte_<A>(x: Array<A>): (y: Array<A>) => boolean;
+export function gte_<A>(x: IArguments, y: IArguments): boolean;
+export function gte_<A>(x: IArguments): (y: IArguments) => boolean;
+export function gte_<A>(x: Ord<A>, y: Ord<A>): boolean;
+export function gte_<A>(x: Ord<A>): (y: Ord<A>) => boolean;
+
+export function min<A>(x: Date, y: Date): Date;
+export function min<A>(x: Date): (y: Date) => Date;
+export function min<A>(x: number, y: number): number;
+export function min<A>(x: number): (y: number) => number;
+export function min<A>(x: string, y: string): string;
+export function min<A>(x: string): (y: string) => string;
+export function min<A>(x: Ord<A>, y: Ord<A>): Ord<A>;
+export function min<A>(x: Ord<A>): (y: Ord<A>) => Ord<A>;
+
+export function max<A>(x: Date, y: Date): Date;
+export function max<A>(x: Date): (y: Date) => Date;
+export function max<A>(x: number, y: number): number;
+export function max<A>(x: number): (y: number) => number;
+export function max<A>(x: string, y: string): string;
+export function max<A>(x: string): (y: string) => string;
+export function max<A>(x: Ord<A>, y: Ord<A>): Ord<A>;
+export function max<A>(x: Ord<A>): (y: Ord<A>) => Ord<A>;
+
+export function id<A>(typeRep: TypeRep): (A) => A;
+export function id<A>(typeRep: TypeRep): Category<A>;
+
+export function concat<A>(x: string, y: string): string;
+export function concat<A>(x: string): (y: string) => string;
+export function concat<A>(x: Array<A>, y: Array<A>): Array<A>;
+export function concat<A>(x: Array<A>): (y: Array<A>) => Array<A>;
+export function concat<A>(x: Semigroup<A>, y: Semigroup<A>): Semigroup<A>;
+export function concat<A>(x: Semigroup<A>): (y: Semigroup<A>) => Semigroup<A>;
+
+export function empty<A>(typeRep: TypeRep): Monoid<A>;
+
+export function map<A, B>(f: (A) => B, xs: Array<A>): Array<B>;
+export function map<A, B>(f: (A) => B): (xs: Array<A>) => Array<B>;
+export function map<A, B>(f: (A) => B, functor: Functor<A>): Functor<B>;
+export function map<A, B>(f: (A) => B): (functor: Functor<A>) => Functor<B>;
+
+export function bimap<A, B, C, D>(f: (A) => B, g: (C) => D, bifunctor: Bifunctor<A, C>): Bifunctor<B, D>;
+export function bimap<A, B, C, D>(f: (A) => B, g: (C) => D): (bifunctor: Bifunctor<A, C>) => Bifunctor<B, D>;
+export function bimap<A, B, C, D>(f: (A) => B): AwaitingTwo<(C) => D, Bifunctor<A, C>, Bifunctor<B, D>>;
```

Comment:
Should it be `S.Left('str')`?

See [TS Playground](http://www.typescriptlang.org/play/index.html#src=interface%20Functor%3CA%3E%20%7B%0A%20%20'fantasy-land%2Fmap'%3A%20%3CB%3E(f%3A%20(a%3A%20A)%20%3D%3E%20B)%20%3D%3E%20Functor%3CB%3E%0A%7D%0A%0Ainterface%20Bifunctor%3CA%2C%20C%3E%20extends%20Functor%3CC%3E%20%7B%0A%20%20'fantasy-land%2Fbimap'%3A%20%3CB%2C%20D%3E(f%3A%20(a%3A%20A)%20%3D%3E%20B%2C%20g%3A%20(c%3A%20C)%20%3D%3E%20D)%20%3D%3E%20Bifunctor%3CB%2C%20D%3E%0A%7D%0A%0Adeclare%20function%20toUpper(s%3A%20string)%3A%20string%3B%0A%0Adeclare%20function%20Left%3CA%2C%20B%3E(value%3A%20A)%3A%20Either%3CA%2C%20B%3E%3B%0A%0Ainterface%20Either%3CA%2C%20B%3E%20%7B%0A%20%20'fantasy-land%2Fequals'%3A%20(other%3A%20Either%3CA%2C%20B%3E)%20%3D%3E%20boolean%0A%20%20'fantasy-land%2Fmap'%3A%20%3CC%3E(f%3A%20(b%3A%20B)%20%3D%3E%20C)%20%3D%3E%20Either%3CA%2C%20C%3E%0A%20%20'fantasy-land%2Fbimap'%3A%20%3CC%2C%20D%3E(f%3A%20(a%3A%20A)%20%3D%3E%20C%2C%20g%3A%20(b%3A%20B)%20%3D%3E%20D)%20%3D%3E%20Either%3CC%2C%20D%3E%0A%20%20'fantasy-land%2Fap'%3A%20%3CC%3E(other%3A%20Either%3CA%2C%20(b%3A%20B)%20%3D%3E%20C%3E)%20%3D%3E%20Either%3CA%2C%20C%3E%0A%20%20'fantasy-land%2Fchain'%3A%20%3CC%3E(f%3A%20(b%3A%20B)%20%3D%3E%20Either%3CA%2C%20C%3E)%20%3D%3E%20Either%3CA%2C%20C%3E%0A%20%20'fantasy-land%2Falt'%3A%20(other%3A%20Either%3CA%2C%20B%3E)%20%3D%3E%20Either%3CA%2C%20B%3E%0A%20%20'fantasy-land%2Freduce'%3A%20%3CC%3E(f%3A%20(c%3A%20C%2C%20b%3A%20B)%20%3D%3E%20C%2C%20x%3A%20C)%20%3D%3E%20C%0A%20%20'fantasy-land%2Ftraverse'%3A%20%3CC%3E(typeRep%3A%20TypeRep%2C%20f%3A%20(b%3A%20B)%20%3D%3E%20Applicative%3CC%3E)%20%3D%3E%20Applicative%3CEither%3CA%2C%20C%3E%3E%0A%7D%0A%0Ainterface%20Apply%3CA%3E%20extends%20Functor%3CA%3E%20%7B%0A%20%20'fantasy-land%2Fap'%3A%20%3CB%3E(other%3A%20Apply%3C(a%3A%20A)%20%3D%3E%20B%3E)%20%3D%3E%20Apply%3CB%3E%0A%7D%0Ainterface%20ApplicativeTypeRep%3CA%3E%20%7B%0A%20%20'fantasy-land%2Fof'%3A%20(value%3A%20A)%20%3D%3E%20Applicative%3CA%3E%0A%7D%0Ainterface%20Applicative%3CA%3E%20extends%20Apply%3CA%3E%20%7B%0A%20%20constructor%3A%20ApplicativeTypeRep%3CA%3E%0A%7D%0A%0Ainterface%20TypeRep%20%7B%7D%0A%0Adeclare%20function%20bimap%3CA%2C%20B%2C%20C%2C%20D%3E(f%3A%20(a%3A%20A)%20%3D%3E%20B%2C%20g%3A%20(c%3A%20C)%20%3D%3E%20D%2C%20bifunctor%3A%20Bifunctor%3CA%2C%20C%3E)%3A%20Bifunctor%3CB%2C%20D%3E%3B%0Adeclare%20function%20bimap%3CA%2C%20B%2C%20C%2C%20D%3E(f%3A%20(a%3A%20A)%20%3D%3E%20B%2C%20g%3A%20(c%3A%20C)%20%3D%3E%20D)%3A%20(bifunctor%3A%20Bifunctor%3CA%2C%20C%3E)%20%3D%3E%20Bifunctor%3CB%2C%20D%3E%3B%0Adeclare%20function%20bimap%3CA%2C%20B%3E(f%3A%20(a%3A%20A)%20%3D%3E%20B)%3A%20%7B%0A%20%20%3CC%2C%20D%3E(x%3A%20(c%3A%20C)%20%3D%3E%20D%2C%20y%3A%20Bifunctor%3CA%2C%20C%3E)%3A%20Bifunctor%3CB%2C%20D%3E%3B%0A%20%20%3CC%2C%20D%3E(x%3A%20(c%3A%20C)%20%3D%3E%20D)%3A%20(y%3A%20Bifunctor%3CA%2C%20C%3E)%20%3D%3E%20Bifunctor%3CB%2C%20D%3E%3B%0A%7D%3B%0A%0Aconst%20result%20%3D%20bimap(toUpper%2C%20Math.sqrt%2C%20Left(42))%0A).